File: graphs/cr8.py
n, m = map(int, input().split())
n += 1
E = [[] for _ in range(n)]

for i in range(m):
    u, v = map(int, input().split())
    E[u].append(v)
    E[v].append(u)

used = [False] * n
allcc = []

def dfs(v: int, cc: list[int]) -> list[int]:
    used[v] = True
    cc.append(v)
    for to in E[v]:
        print("E[v]:", E[v])
        if used[to] == False:
            dfs(to, cc)
    return cc

for i in range(1, n):
    if used[i] == False:
        allcc.append(dfs(i, []))


# print(len(allcc))

# for elem in allcc:
#     print(len(elem))
#     print(*elem)

File: graphs/cr8.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> dfs(int v, vector<vector<int>>& E ,vector<bool>& used, vector<int>& cc) {
    used[v] = true;
    cc.push_back(v);
    // for (int i = 1; i < E[v].size(); i++) {
    //     cout << i << endl;
    //     if (!used[v]) dfs(E[v][i], E, used, cc);
    // }
    for (int to : E[v]) {
        // cout << "TO:";
        if (!used[to]) dfs(to, E, used, cc);
    }


    return cc;
    
}


int main () {
    ios::sync_with_stdio(NULL);
    int n, m;
    cin >> n >> m;
    n++;
    // unordered_set<int> us ();
    vector<vector<int>> E(n);

    while(m--) {
        int u, v;
        cin >> u >> v;
        E[v].push_back(u);
        E[u].push_back(v);
    }

    vector<bool> used(n);
    for (int i = 0; i < n; i++) {
        used[i] = false;
    }

    vector<vector<int>> allcc;

    for (int i = 1; i < n; i++) {
        vector<int> cc;
        if (!used[i]) {
            dfs(i, E, used, cc);
            allcc.push_back(cc);
        }
    }

    cout << allcc.size() << "\n";

    for (vector<int> vec : allcc) {
        cout << vec.size() << "\n";
        for (int i = 0; i < vec.size();i++) cout << vec[i] << " ";
        cout << endl;
    }
    
}